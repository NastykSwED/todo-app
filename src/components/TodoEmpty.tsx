export const TodoEmpty = () => {
  return (
    <div className="re-TodoEmpty">No Todos</div>
  )
}
