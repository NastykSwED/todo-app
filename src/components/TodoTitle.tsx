export const TodoTitle = () => {
	return (
		<h1 className='re-TodoTitle'>
			My todos
		</h1>
	);
};
