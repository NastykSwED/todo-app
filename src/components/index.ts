export * from './TodoAdd';
export * from './TodoList';
export * from './TodoItem';
export * from './TodoTitle';
export * from './TodoEmpty'