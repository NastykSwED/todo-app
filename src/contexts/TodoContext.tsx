import { createContext } from 'react';

export const TodoContext = createContext<TodoContextProps>(
	{} as TodoContextProps
);
