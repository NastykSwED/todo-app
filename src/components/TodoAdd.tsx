import { useState, useContext } from 'react';

import { TodoContext } from '../contexts/TodoContext';

export const TodoAdd = () => {
	const { handleAddTodo } = useContext(TodoContext);

	const [nameInputValue, setNameInputValue] = useState('');
	const [descriptionInputValue, setDescriptionInputValue] = useState('');

	const handleChangeName = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { target } = e;
		const newNameInputValue = target.value;

		setNameInputValue(newNameInputValue);
	};

	const handleChangeDescription = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { target } = e;
		const newDescriptionInputValue = target.value;

		setDescriptionInputValue(newDescriptionInputValue);
	};

	return (
		<div className='re-TodoAdd'>
			<div className='re-TodoAdd-wrapInputs'>
				<label className='re-TodoAdd-label' htmlFor=''>
					Name
					<input
						placeholder='Todo Name'
						className='re-TodoAdd-input'
						value={nameInputValue}
						onChange={handleChangeName}
						type='text'
					/>
				</label>

				<label className='re-TodoAdd-label' htmlFor=''>
					Description
					<input
						placeholder='Todo Description'
						className='re-TodoAdd-input'
						value={descriptionInputValue}
						onChange={handleChangeDescription}
						type='text'
					/>
				</label>
			</div>

			<button
				onClick={() => {
					handleAddTodo(nameInputValue, descriptionInputValue);

					setNameInputValue('');
					setDescriptionInputValue('');
				}}
				className='re-TodoAdd-button'
			>
				Add Todo
			</button>
		</div>
	);
};
