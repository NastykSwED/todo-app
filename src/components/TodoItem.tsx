import { useContext, useState } from 'react';
import { TodoContext } from '../contexts/TodoContext';

interface TodoItemProps {
	todo: TodoProps;
}

export const TodoItem = (props: TodoItemProps) => {
	const { todo } = props;

	const { name, description } = todo;

	const [isCompleted, setIsCompleted] = useState(false);

	const { handleDeleteTodo, handleUpdateTodo } = useContext(TodoContext);

	return (
		<div className='re-TodoItem'>
			<div className='re-TodoItem-wrapBodyText'>
				<h2
					className={`re-TodoItem-title ${isCompleted ? 'isCompleted' : ''} `}
				>
					{name}
				</h2>
				<p
					className={`re-TodoItem-description ${
						isCompleted ? 'isCompleted isCompleted--text' : ''
					} `}
				>
					{description}
				</p>
			</div>

			<div className='re-TodoItem-wrapButtons'>
				<button
					onClick={() => {
						setIsCompleted(!isCompleted);

						handleUpdateTodo(todo);
					}}
					className={`re-TodoItem-button TodoItem-button--completed`}
				>
					Complete
				</button>
				<button
					onClick={() => handleDeleteTodo(todo)}
					className='re-TodoItem-button TodoItem-button--deleted'
				>
					Delete
				</button>
			</div>
		</div>
	);
};
