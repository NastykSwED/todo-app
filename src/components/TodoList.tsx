import { TodoEmpty, TodoItem } from './';

import { useContext } from 'react';

import { TodoContext } from '../contexts/TodoContext';

export const TodoList = () => {
	const { todos } = useContext(TodoContext);

	return (
		<div className='re-TodoList'>
			{todos.length > 0 ? (
				todos.map(todo => <TodoItem key={todo.id} todo={todo} />)
			) : (
				<TodoEmpty />
			)}
		</div>
	);
};
