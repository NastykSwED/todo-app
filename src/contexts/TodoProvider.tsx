import { TodoContext } from './TodoContext';

import { useTodo } from '../hooks/useTodo';

import { useEffect } from 'react';

interface TodoProviderProps {
	children: React.ReactNode;
}

export const TodoProvider = (props: TodoProviderProps) => {
	const { children } = props;

	const { todos, handleAddTodo, handleUpdateTodo, handleDeleteTodo } =
		useTodo();

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const value = {
		todos,
		handleAddTodo,
		handleDeleteTodo,
		handleUpdateTodo,
	};

	return <TodoContext.Provider value={value}>{children}</TodoContext.Provider>;
};
