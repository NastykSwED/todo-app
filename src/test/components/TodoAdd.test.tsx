import { describe, expect, it, vi } from 'vitest';

import { screen, render, fireEvent } from '@testing-library/react';

import { TodoAdd } from '../../components';

import { TodoContext } from '../../contexts/TodoContext';

describe('Testing TodoAdd Component', () => {
	const value = {
		todos: [],
		handleAddTodo: vi.fn(),
		handleDeleteTodo: vi.fn(),
		handleUpdateTodo: vi.fn(),
	};

	it('Should match the snapshot', () => {
		const { container } = render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		expect(container).toMatchSnapshot();
	});

	it('The todo name input should be empty', () => {
		render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		const todoNameInput = screen.getByPlaceholderText(
			'Todo Name'
		) as HTMLInputElement;
		const emptyValue = '';

		expect(todoNameInput.value).toBe(emptyValue);
	});

	it('A new value should be added to the todo name input', () => {
		render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		const todoNameInput = screen.getByPlaceholderText(
			'Todo Name'
		) as HTMLInputElement;

		const newInputValue = 'Learn ReactJS';

		fireEvent.change(todoNameInput, {
			target: {
				value: newInputValue,
			},
		});

		expect(todoNameInput.value).toBe(newInputValue);
	});

	it('The todo description input should be empty', () => {
		render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		const todoDescriptionInput = screen.getByPlaceholderText(
			'Todo Description'
		) as HTMLInputElement;
		const emptyValue = '';

		expect(todoDescriptionInput.value).toBe(emptyValue);
	});

	it('A new value should be added to the todo description input', () => {
		render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		const todoDescriptionInput = screen.getByPlaceholderText(
			'Todo Name'
		) as HTMLInputElement;

		const newInputValue = 'I want to learn this amazing framework';

		fireEvent.change(todoDescriptionInput, {
			target: {
				value: newInputValue,
			},
		});

		expect(todoDescriptionInput.value).toBe(newInputValue);
	});

	it('The todo name input and the todo description input should be enter when we add a todo', () => {
		render(
			<TodoContext.Provider value={value}>
				<TodoAdd></TodoAdd>
			</TodoContext.Provider>
		);

		const todoNameInput = screen.getByPlaceholderText(
			'Todo Name'
		) as HTMLInputElement;

		const todoDescriptionInput = screen.getByPlaceholderText(
			'Todo Description'
		) as HTMLInputElement;

		const emptyValue = '';

		const newNameInputValue = 'Learn ReactJS';
		const newDescriptionInputValue = 'I want to learn this amazing framework';

		fireEvent.change(todoNameInput, {
			target: {
				value: newNameInputValue,
			},
		});

		fireEvent.change(todoDescriptionInput, {
			target: {
				value: newDescriptionInputValue,
			},
		});

		fireEvent.click(screen.getByRole('button'));

		expect(todoNameInput.value).toBe(emptyValue);

		expect(todoDescriptionInput.value).toBe(emptyValue);
	});
});
