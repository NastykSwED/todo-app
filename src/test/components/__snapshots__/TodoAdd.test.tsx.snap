// Vitest Snapshot v1

exports[`Testing TodoAdd Component > Should match the snapshot 1`] = `
<div>
  <div
    class="re-TodoAdd"
  >
    <div
      class="re-TodoAdd-wrapInputs"
    >
      <label
        class="re-TodoAdd-label"
        for=""
      >
        Name
        <input
          class="re-TodoAdd-input"
          placeholder="Todo Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="re-TodoAdd-label"
        for=""
      >
        Description
        <input
          class="re-TodoAdd-input"
          placeholder="Todo Description"
          type="text"
          value=""
        />
      </label>
    </div>
    <button
      class="re-TodoAdd-button"
    >
      Add Todo
    </button>
  </div>
</div>
`;
